agent/rpc/peering/replication.go

```diff
@@ -282,8 +288,136 @@ func (s *Service) handleUpsertService(
 		}
 	}
 
-	// TODO(peering): cleanup and deregister existing data that is now missing safely somehow
+	//
+	// Now that the data received has been stored in the state store, the rest of this
+	// function is responsible for cleaning up data in the catalog that wasn't in the snapshot.
+	//
+
+	// nodeCheckTuple uniquely identifies a node check in the catalog.
+	// The partition is not needed because we are only operating on one partition's catalog.
+	type nodeCheckTuple struct {
+		checkID types.CheckID
+		node    string
+	}
+
+	var (
+		// unusedNodes tracks node names that were not present in the latest response.
+		// Missing nodes are not assumed to be deleted because there may be other service names
+		// registered on them.
+		// Inside we also track a map of node checks associated with the node.
+		unusedNodes = make(map[types.NodeID]struct{})
+
+		// deletedNodeChecks tracks node checks that were not present in the latest response.
+		// A single node check will be attached to all service instances of a node, so this
+		// deduplication prevents issuing multiple deregistrations for a single check.
+		deletedNodeChecks = make(map[nodeCheckTuple]struct{})
```

Comment:
Conveniently after i re-activate mesh-gateway-only-mode there will be no node checks replicated to worry about since the checks on a service will be flattened into a single summary check per service instance.

Reply:
Right, unless we wanted to squash those checks as well.